docs/specifications/url.rst

```diff
@@ -0,0 +1,112 @@
+These are not inherently tied to the Foolscap protocol itself.
+In particular they are beneficial to :doc:`../proposed/http-storage-node-protocol` which uses HTTP instead of Foolscap.
+It is conceivable they will also be used with WebSockets at some point as well.
+
+Continuing to refer to these URLs as fURLs when they are being used for other protocols may cause confusion.
+Therefore,
+this document coins the name **NURL** for these URLs.
+This can be considered to expand to "**N**\ ew URLs" or "Authe\ **N**\ ticating URLs" or "Authorizi\ **N**\ g URLs" as the reader prefers.
+
+Syntax
+------
+
+The EBNF for a NURL is as follows::
+
+  nurl         = scheme, hash, "@", net-loc-list, "/", swiss-number, [ version1 ]
+
+  scheme       = "pb://"
+
+  hash         = unreserved
+
+  net-loc-list = net-loc, [ { ",", net-loc } ]
+  net-loc      = hostname, [ ":" port ]
+  hostname     = domain | IPv4address | IPv6address
+
+  swiss-number = segment
+
+  version1     = "#v=1"
+
+See https://tools.ietf.org/html/rfc3986#section-3.3 for the definition of ``segment``.
+See https://tools.ietf.org/html/rfc2396#appendix-A for the definition of ``unreserved``.
+See https://tools.ietf.org/html/draft-main-ipaddr-text-rep-02#section-3.1 for the definition of ``IPv4address``.
+See https://tools.ietf.org/html/draft-main-ipaddr-text-rep-02#section-3.2 for the definition of ``IPv6address``.
+See https://tools.ietf.org/html/rfc1035#section-2.3.1 for the definition of ``domain``.
+
+Versions
+--------
+
+Though all NURLs are syntactically compatible some semantic differences are allowed.
+These differences are separated into distinct versions.
+
+Version 0
+---------
+
+A Foolscap fURL is considered the canonical definition of a version 0 NURL.
+Notably,
+the hash component is defined as the base32-encoded SHA1 hash of the DER form of an x509v3 certificate.
+A version 0 NURL is identified by the absence of the ``v=1`` fragment.
+
+Version 1
+---------
+
+The hash component of a version 1 NURL differs in three ways from the prior version.
+
+1. The hash function used is SHA3-224 instead of SHA1.
+   The security of SHA1 `continues to be eroded`_.
+   Contrariwise SHA3 is currently the most recent addition to the SHA family by NIST.
+   The 224 bit instance is chosen to keep the output short and because it offers greater collision resistance than SHA1 was thought to offer even at its inception
+   (prior to security research showing actual collision resistance is lower).
+2. The hash is computed over the certificate's SPKI instead of the whole certificate.
+   This allows certificate re-generation so long as the public key remains the same.
+   This is useful to allow contact information to be updated or extension of validity period.
+   Use of an SPKI hash has also been `explored by the web community`_ during its flirtation with using it for HTTPS certificate pinning
+   (though this is now largely abandoned).
```

Comment:
Might be worth pointing mentioning here what the disposition of anything else in the certificate is; it's not "authenticated by the NURL" directly .. but do we ignore it? Fixed strings? *Should* clients ignore other things in the certificate?

Reply:
Either "ignore" or "fixed values" seem like they're alright to me.  Even a variation on "ignore" of "only display for informational/entertainment purposes" is maybe fine?  That does start to get into a more dangerous place, I guess, since maybe whoever you're displaying it to does something other than consume it as entertainment, despite any warnings you provide.

I put some language in here warning about the lack of validation for anything apart from the SPKI hash, maybe that's sufficient for now / for this document?

Also one thing I realized as writing this is this area also isn't so different from Foolscap.  I can put anything I want in a cert and hand it to Foolscap.  The only difference is that the values are frozen from that point on, whereas with NURLv2 I can change them whenever I want.  But either way the values are ultimately under my direct and total control at some point in time.